lib/cog/command/user_permissions_cache.ex

```diff
@@ -83,16 +83,38 @@ defmodule Cog.Command.UserPermissionsCache do
 
   defp fetch_and_cache({adapter, username}=key, state) do
     Logger.info("Cache miss for #{inspect key}")
-    case Repo.one(Queries.User.for_handle(username, adapter)) do
+
+    with {:ok, external_user} <- fetch_user_from_adapter(adapter, username),
+         {:ok, internal_user} <- fetch_user_from_database(adapter, external_user.id) do
+      perms = Cog.Models.User.all_permissions(internal_user)
+      value = {internal_user, perms}
+      expiry = Cog.Time.now() + state.ttl
+      :ets.insert(@ets_table, {key, value, expiry})
+      {:ok, value}
+    end
+  end
+
+  def fetch_user_from_adapter(adapter, handle) do
+    user = with {:ok, adapter_module} <- Cog.adapter_module(adapter),
+                do: adapter_module.lookup_user(handle: handle)
+
+    case user do
+      {:ok, user} ->
+        {:ok, user}
+      {:error, _} ->
+        {:error, :not_found}
+    end
+  end
```

Comment:
The reason I used with was to avoid duplicating the match that swaps out the error for the more generic `{:error, :not_found}`.

The code would be: 

```
case Cog.adapter_module(adapter) do
  {:ok, mod} ->
    case mod.lookup_user(handle: handle) do
      {:ok, user} ->
        {:ok, user}
      {:error, _} ->
        {:error, :not_found}
    end
  error ->
    {:error, _} ->
      {:error, :not_found}
end
```

It is a tiny bit more clear, but I think the `with` version is more maintainable.